データの保存先をPostgreSQLに変更した

nakamu-kazu222 opened the pull request:
データの保存先をJSONファイルからPostgreSQLに変更した
データの保存には、`データベース名：sinatra_memo` が必要になる

nakamu-kazu222 committed:
データの保存先をPostgreSQLに変更した

nakamu-kazu222 committed:
シンボルに統一した

nakamu-kazu222 committed:
Rubocop警告でhelper内の行数が多かったため、コードを簡略化した

nakamu-kazu222 committed:
DB保存用の手順を記載した

maedana commented on memo_app.rb:
```diff
@@ -2,15 +2,16 @@
+configure do
+  connection = PG.connect(dbname: 'sinatra_memo', user: 'postgres')
+  connection.exec("SELECT * FROM information_schema.tables WHERE table_name = 'memos'")
+end
```
このconfigureブロック何をしているものなのでしょう? 意図がわからないので説明して欲しいです。

nakamu-kazu222 replied:
アプリケーションが開始された時に最初に実行され、PostgreSQLのデータベース名、ユーザー名、テーブル名が正しく設定されているか確認し、設定されていない場合はアプリケーションを実行することができないようにしました

maedana replied:
なるほど、意図としてはわかりました。しかし、テーブル名が正しくても、テーブル内のカラムが違っていれば意味がないのではでしょうか?

また何かがおかしいときは`pg` gemの例外が出るだけですよね。それって結局メモ一覧画面を開いたりすときに似たようなエラー出るでしょうし、起動時に別途実施するメリットがいまいちよくわからなかったです。必要なDBが用意されてないときに自動で作成してくれる、とかならまだわかりますが、今回の課題においては正しいデータベースがセットアップされている前提でいいかなーと思います。

nakamu-kazu222 replied:
> メモ一覧画面を開いたりすときに似たようなエラー出るでしょうし、起動時に別途実施するメリットがいまいちよくわからなかったです
必要なDBが用意されてないときに自動で作成してくれる、とかならまだわかりますが

おっしゃる通りで、メリットがないと理解しました
たしかに、DBの自動生成でしたらconfigureブロックで最初に行えばいいですね

正しいデータベースがセットアップされている前提で課題を進めるので、configureブロックの処理は省略します

maedana commented on memo_app.rb:
```diff
@@ -19,29 +20,38 @@ def make_id
   def save_memo(memo)
-    File.open(memo_data_json_file_path(memo[:id]), 'w') { |file| JSON.dump(memo, file) }
+    db_connection do |connection|
+      if memo[:id].nil? || get_memo(memo[:id]).nil?
```
すでにコネクション貼っていてさらに`get_memo`でコネクション貼ることになり、2つコネクション使うことになりますよね。1コネクションで済むようにしたいかなと思いますけどいかがでしょうか?

注意点にも
> データベースに接続するのは重い処理です。1リクエストで何度も接続しないように工夫しましょう。

のようにありますねー。

nakamu-kazu222 replied:
おっしゃる通り2つコネクションを使用していました
1リクエストで1コネクションになるように修正します

maedana commented on memo_app.rb:
```diff
@@ -69,7 +79,7 @@ def save_memo(memo)
 post '/memos' do
   memo = {
-    id: make_id,
+    id: make_id.to_sym,
```
to_symの必要性がよくわかりませんでした。必要な理由があれば教えてください :pray:

nakamu-kazu222 replied:
memo内のキーをシンボルで統一しようと考えていました
しかし、idキーはすでにシンボルになっており、`to_sym`が必要ないことがわかりましたので削除しました

nakamu-kazu222 committed:
1リクエストで1コネクションになるように変更した

重い処理であるデータベースに接続を軽減するために変更した

nakamu-kazu222 committed:
idキーはすでにシンボルであるため、不要なメソッドを削除した

maedana commented on memo_app.rb:
```diff
@@ -50,45 +56,53 @@ def save_memo(memo)
+  db_connection do |connection|
+    @memo = get_memo(params[:id], connection)
+    if @memo.nil?
+      erb :not_found_error
```
ここに限らずですけど、erbのレンダリングやredirectまでdb接続しっぱなしである必要ないですよね。db接続の回数を少なくしつつ、さらに接続している時間も最低限になるようにしましょうー。

nakamu-kazu222 committed:
DB接続を最小限にした

DB に接続してレコード情報を取得する処理をしたら、DB 接続は必要なくなる
そのため、db_connectionの処理を終了させ DB 接続を解除するように変更した

nakamu-kazu222 committed:
正しいデータベースがセットアップされている前提なので、configureブロックの処理は省略した

nakamu-kazu222 committed:
SELECTで1件取得しようとしていることを明示的にした
